### PR TITLE
Create User obj when signing in via LDAP

### PR DESCRIPTION
**Pull request by JeffreyThiessen**
tbd below
## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [ ] Tests added (or description of how to test) for any new features.
* [ ] User documentation updated for UI or technical changes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@
 * [Developer]: Fixed deleting a project with project subscriptions. See [PR 1348](https://github.com/phac-nml/irida/pull/1348)
 * [Developer]: Updated OAuth2 implementation to use Spring Security 5 OAuth2 libraries. See [PR 1339](https://github.com/phac-nml/irida/pull/1339)
 * [Developer]: Add identifier to project drop-downs. See [PR 1352](https://github.com/phac-nml/irida/pull/1352)
+* [ALL]: Added LDAP/ADLDAP support.
```
Comment (ericenns) [inline CHANGELOG.md:21]: Is this changelog comment going to be updated to include PR reference in the main branch (i.e. ldap)

File: build.gradle.kts [hunk 2]
```diff
@@ -214,6 +214,9 @@ dependencies {
     testImplementation("org.seleniumhq.selenium:selenium-support:3.141.59")
     testImplementation("org.seleniumhq.selenium:selenium-chrome-driver:3.141.59")
     testImplementation("org.mockftpserver:MockFtpServer:2.6")
+    testImplementation("org.springframework.ldap:spring-ldap-test")
+    implementation("org.springframework.boot:spring-boot-starter-data-ldap")
```
Comment (ericenns) [inline build.gradle.kts:218]: Is this dependency actually required (i.e. `org.springframework.boot:spring-boot-starter-data-ldap`) if so can it be moved up the `spring-ldap-core` dependency.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+    /**
+     * Fetch data from LDAP/ADLDAP and create a map that aligns with {@link User}
+     * @param dirContextOperations LDAP/ADLDAP context
+     * @return Map<String {@link User} property, String LDAP/ADLDAP field value>
+     */
+    private Map<String, String> getLdapFields(DirContextOperations dirContextOperations) {
+        Map<String, String> map = new HashMap<>();
+        map.put(iridaUserFieldEmail, getAttribute(dirContextOperations, userInfoEmail, true, ""));
+        map.put(iridaUserFieldFirstName, getAttribute(dirContextOperations, userInfoFirstName, false, "FirstName"));
```
Comment (ericenns) [inline src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java:213]: Should we really have a fallback for `firstName` and `lastName`, I lean towards no since this would be very confusing to see the list of users in IRIDA since these won't be modifiable.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+            }
+        }
+        return res;
+    }
+
+    /**
+     * Generates a random password that adheres to our password requirements
+     * @return String
+     */
+    public String generateCommonLangPassword() {
```
Comment (ericenns) [inline src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java:305]: Instead of having to generate a random password wouldn't it make more sense to create another entity i.e. `DomainUser` that allows Null for password field. Something like how we have `CoverageQCEntry` that is on top of `QCEntry` and has the `DTYPE` column. That might complicate password checking for non `DomainUser` accounts as we would not have the db NULL constraint. I guess this could be tackled in your other PR where you allow local and domain users.

**Review by ericenns — CHANGES_REQUESTED** [review]
Just some minor comments for now. I still need to run the tests locally though.

File: CHANGELOG.md (hunk 1 above)
Comment (JeffreyThiessen) [reply CHANGELOG.md:21]: Yes that is the plan. Once we have all the features in the `ldap` branch and it's ready for a PR to dev it will be added.

File: build.gradle.kts (hunk 2 above)
Comment (JeffreyThiessen) [reply build.gradle.kts]: Yep, having it be a required (not a test) dependency was needed to get it to start with the tests correctly. Moved now.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java (hunk 3 above)
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java]: Good catch. I have changed the name fields fields to be required now.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java (hunk 4 above)
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java:305]: That's likely the better way to do it yeah. I'll work on that as part of the next PR as this one already is a bit bloated.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/relational/auditing/UserRevListener.java [hunk 5]
```diff
@@ -60,14 +68,32 @@ public void setApplicationContext(ApplicationContext applicationContext) throws
     public void initialize() {
         urepo = applicationContext.getBean(UserRepository.class);
         clientRepo = applicationContext.getBean(IridaClientDetailsRepository.class);
+        try { // Only gets the context mapper if we are in ldap mode
+            iridaUserDetailsContextMapper = applicationContext.getBean(IridaUserDetailsContextMapper.class);
+        } catch (NoSuchBeanDefinitionException ignored){}
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/relational/auditing/UserRevListener.java]: Seems like an edge case that would get caught before they try to commit a user revision, but doesn't hurt to be safe. Added.

**Comment by JeffreyThiessen** [comment]
How to manually test this branch

1. Download apache directory studio from here https://directory.apache.org/studio/downloads.html
2. https://docs.craftercms.org/en/3.0/developers/cook-books/how-tos/setting-up-an-ldap-server-for-dev.html

The  tutorial is not exactly what I used but it does explain stuff pretty well. Skip past the "crafter" stuff at the beginning and then it explains how to set up apache directory studio and make a connection so you can load a .ldif file.

Here's some data to save as an .ldif file to populate your ldap server
```
dn: dc=springframework,dc=org
objectclass: top
objectclass: domain
objectclass: extensibleObject
dc: springframework

dn: ou=groups,dc=springframework,dc=org
objectclass: top
objectclass: organizationalUnit
ou: groups

dn: ou=subgroups,ou=groups,dc=springframework,dc=org
objectclass: top
objectclass: organizationalUnit
ou: subgroups

dn: ou=people,dc=springframework,dc=org
objectclass: top
objectclass: organizationalUnit
ou: people

dn: uid=jwick,ou=people,dc=springframework,dc=org
objectclass: top
objectclass: person
objectclass: organizationalPerson
objectclass: inetOrgPerson
cn: John Wick
sn: Wick
uid: jwick
givenName: John
mail: jwick@example.com
telephoneNumber: 169-637-3314
userPassword: Password1!

dn: uid=jdoe,ou=people,dc=springframework,dc=org
objectclass: top
objectclass: person
objectclass: organizationalPerson
objectclass: inetOrgPerson
cn: Jane Doe
sn: Doe
description: 19650324000000Z
employeeNumber: 12
givenName: Jane
mail: jane@example.com
telephoneNumber: 907-547-9114
uid: jdoe
userPassword: Password1
```

3. Change or override the properties in resources/ca/corefacility/bioinformatics/irida/config/authentication.properties

```
# ldap properties
irida.administrative.authentication.mode=ldap

# LDAP config
irida.administrative.authentication.ldap.url=ldap://localhost:18880/
#irida.administrative.authentication.ldap.base=dc=example,dc=com
irida.administrative.authentication.ldap.base=dc=springframework,dc=org
irida.administrative.authentication.ldap.userdn=uid=admin,ou=system
irida.administrative.authentication.ldap.password=secret
irida.administrative.authentication.ldap.userdn_search_patterns=uid={0},ou=people
# Optional Argument. change this to 'follow' to handle referrals. See AbstractContextSource javadocs for more info
irida.administrative.authentication.ldap.set_referral=ignore

# Required fields for LDAP and ADLDAP to fetch required information when creating new users
irida.administrative.authentication.ldap.userInfoEmail=mail
irida.administrative.authentication.ldap.userInfoFirstName=givenname
irida.administrative.authentication.ldap.userInfoLastName=sn
irida.administrative.authentication.ldap.userInfoPhoneNumber=telephonenumber
```

4. Change a users password, because we want it to have some encryption on the ldap side
5. Sign in using the user uid and the password
6. If the user does not already exist in the local irida db, it will create the new user.
7. If you change a users information on ldap (givenname, sn, mail, telephonenumber) and sign in, the local user object will update.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+     * Fetch data from LDAP/ADLDAP and create a map that aligns with {@link User}
+     * @param dirContextOperations LDAP/ADLDAP context
+     * @return Map<String {@link User} property, String LDAP/ADLDAP field value>
+     */
+    private Map<String, String> getLdapFields(DirContextOperations dirContextOperations) {
+        Map<String, String> map = new HashMap<>();
+        map.put(iridaUserFieldEmail, getAttribute(dirContextOperations, userInfoEmail, true, ""));
+        map.put(iridaUserFieldFirstName, getAttribute(dirContextOperations, userInfoFirstName, true, ""));
+        map.put(iridaUserFieldLastName, getAttribute(dirContextOperations, userInfoLastName, true, ""));
+        map.put(iridaUserFieldPhoneNumber, getAttribute(dirContextOperations, userInfoPhoneNumber, false, "0000"));
```
Comment (ericenns) [inline src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java:215]: since phone number is no longer required could this default be changed to Null.
